File: Tema 3 - Web API/PetShelter/PetShelter.Domain/Services/FundraiserService.cs
using System;
using PetShelter.DataAccessLayer.Models;
using PetShelter.DataAccessLayer.Repository;
using PetShelter.Domain.Extensions.DomainModel;
using PetShelter.Domain.Extensions.DataAccess;

namespace PetShelter.Domain.Services
{
    public class FundraiserService : IFundraiserService
    {
        private readonly IFundraiserRepository _fundraiserRepository;
        private readonly IDonationRepository _donationRepository;
        private readonly IPersonRepository _personRepository;

        public FundraiserService(IFundraiserRepository fundraiserRepository, IDonationRepository donationRepository, IPersonRepository personRepository)
        {
            _fundraiserRepository = fundraiserRepository;
            _donationRepository = donationRepository;
            _personRepository = personRepository;
        }

        public async Task<int> CreateFundraiser(Person owner, Fundraiser fundraiser)
        {
            var person = await _personRepository.GetOrAddPersonAsync(owner.FromDomainModel());
            var fundraiserAdded = new DataAccessLayer.Models.Fundraiser(fundraiser.Id, fundraiser.Name, fundraiser.GoalValue, fundraiser.DueDate)
            {
                Owner = person
            };
            await _fundraiserRepository.Add(fundraiserAdded);
            return fundraiserAdded.Id;
        }


        public async Task DeleteFundraiserAsync(int fundraiserId)
        {
            var fundraiser = await _fundraiserRepository.GetById(fundraiserId);

            await _fundraiserRepository.Delete(fundraiser);
        }

        public async Task<IReadOnlyCollection<Fundraiser>> GetAllFundraisers()
        {
            var fundraisers = await _fundraiserRepository.GetAll();

            return fundraisers.Select(x => x.toDomainModel()).ToArray();
        }

        public async Task<Fundraiser> GetFundraiserAsync(int fundraiserId)
        {
            var fundraiser = await _fundraiserRepository.GetById(fundraiserId);

            if (fundraiser == null)
            {
                return null;
            }

            fundraiser.Owner = await _personRepository.GetById(fundraiser.OwnerId);
            fundraiser.Donations = await _donationRepository.GetAllDonorsForSpecificFundraiser(fundraiserId);
            return fundraiser.toDomainModel();
        }

        public async Task DonateToFundraiser(int fundraiserId, Donation donation)
        {
            var person = await _personRepository.GetOrAddPersonAsync(donation.Donor.FromDomainModel());
            donation.DonorId = person.Id;
            var fundraiser = await _fundraiserRepository.GetById(fundraiserId);
            fundraiser.CurrentlyDonationAmount += donation.Amount;
            if (fundraiser.CurrentlyDonationAmount > fundraiser.GoalValue)
            {
                fundraiser.Status = "Closed";
            }
            await _fundraiserRepository.Update(fundraiser);
            await _donationRepository.Add(new DataAccessLayer.Models.Donation(donation.Amount, donation.DonorId, fundraiserId));
        }
    }
}

File: Tema 6 - React/src/Services/FundraiserService.ts
import { Fundraiser, Status } from '../Models/Fundraiser';
import { Person } from '../Models/Person';
import type Donation from '../Models/Donation';
import axios from 'axios';
export class FundraiserService {
  private apiUrl: string = 'https://localhost:7075';

  public async getAll(): Promise<Fundraiser[]> {
    try {
      const response = await axios.get(this.apiUrl + '/Fundraiser');
      return response?.data;
    } catch (e) {
      console.log(e);
    }

    return [];
  }

  public async getOne(id: number): Promise<Fundraiser | null> {
    try {
      const response = await axios.get(this.apiUrl + `/Fundraiser/${id}`);
      const objectReceived: IdentifiableFundraiserDTO = response?.data;
      const owner = new Person(
        objectReceived.owner.name,
        objectReceived.owner.idNumber,
        objectReceived.owner.dateOfBirth,
      );
      const donors: Person[] = [];
      objectReceived.donors.forEach((donor) => {
        donors.push(new Person(donor.name, donor.idNumber,objectReceived.owner.dateOfBirth,));
      });

      return new Fundraiser(
        objectReceived.name,
        objectReceived.goalValue,
        objectReceived.dueDate,
        owner,
        objectReceived.id,
        objectReceived.creationDate,
        objectReceived.status,
        objectReceived.currentlyRaisedAmount,
        donors,
      );
    } catch (e) {
      console.log(e);
    }
    return null;
  }

  public async createFundraiser(fundraiser: Fundraiser): Promise<number|null> {
    try {
       const response=await axios.post(this.apiUrl + '/Fundraiser',fundraiser);
        return response?.data;
    } catch (e) {
      console.log(e);
    }
    return null;
  }

  public async donateToFundraiser(fundraiserId:number,donation:Donation):Promise<boolean>{
    try {
        await axios.post(this.apiUrl + `/Fundraiser/${fundraiserId}/donate`,donation);
        return true;
     } catch (e) {
       console.log(e);
     }
     return false;
  }
}

interface FundraiserDTO {
  name: string;
  goalValue: number;
  currentlyRaisedAmount: number;
  creationDate: Date;
  dueDate: Date;
  status: Status;
}

interface IdentifiableFundraiserDTO extends FundraiserDTO {
  id: number;
  owner: Person;
  donors: Person[];
}